### PR TITLE
Merge STANDARD_MODE_SAFE_RECOVERY and MAV_STANDARD_MODE_RETURN_HOME

### PR DESCRIPTION
**Pull request by hamishwillee**
This merges the standard modes MAV_STANDARD_MODE_RETURN_HOME and STANDARD_MODE_SAFE_RECOVERY  into one mode, and renumbers the mode enum.

The original thinking was that it was useful for a user to know whether the vehicle is performing a smart return or a simple/direct RTL. However generally the user is the person who will have done any off-standard configuration, so they will know the expected behaviour. This won't change unless they change it, so telling them they are doing a smart return or an RTL isn't going to be telling them anything they wouldn't know already.

In addition, discussion during prototyping indicates this may be confusing for users. For example, if they set a smart RTL but the vehicle is configured to RTL then it is unclear whether the vehicle should reject a command to set smart recovery mode, or accept it but display the return mode.

The proposal therefore is to have just one return mode that tells the user they are returning to a safe location.

@auturgy @bkueng @julianoes Does this make sense to you? FYI @peterbarker

### CODE REVIEW

**Review by bkueng — APPROVED** [review]
Makes sense to me. Thanks for checking.

**Comment by hamishwillee** [comment]
@auturgy Did you want to sanity check this?

**Comment by hamishwillee** [comment]
OK, this was agreed in mav call, and I have just re-reviewed it. Since it captures intent and passes tests I am going to merge it.

**Comment by hamishwillee** [comment]
The API is WIP. We can break at will @dakejahl.

I guess we could merge these two and leave the other ids unchanged.

**Comment by hamishwillee** [comment]
@dakejahl At this point I am assuming QGC will do a recompile

**Comment by dakejahl** [comment]
@hamishwillee I wonder if we should keep calling it "return" or do I just gotta get used to saying "it's safe recoverying!" :laughing: 

Made a PR into QGC and now the toolbar will show "Safe Recovery" instead of "Return"
https://github.com/mavlink/qgroundcontrol/pull/12276/files#diff-02c4f8918b12c7ca65f7f4ad0b374d0d930aa599f84a8a34d65789d4f9cc4564R59

**Comment by hamishwillee** [comment]
@dakejahl FYI, not ignoring you - went on holiday :-)

Naming things is a fractious business. I'd prefer we keep it as Safe Recovery as you have done. Right now it feels wrong because we're use to saying RTL, but this is a better name because it captures that we aren't necessarily returning anywhere. I hope we'll all get used to it.

**Comment by MaEtUgR** [comment]
> and renumbers the mode enum
> We can break at will

Just a heads-up about the major impact of this breaking change: many users myself included are repeatedly running incompatible ground stations and PX4 versions, leading to incorrect mode displays and unintended actions when e.g. "Return" is pressed. It’s causing significant confusion. While I understand this might not have been predictable, assuming the enum wasn’t used despite QGC having an implementation was incorrect. In the future, it would be great to avoid renumbering enums if possible.

**Comment by hamishwillee** [comment]
@MaEtUgR yes, we do want to avoid making breaking changes. I'm sorry that they happened and I don't want them to happen again.

BUT we created development.xml to allow exactly this kind of thing - we protect the compatibility once things are in common/standard.xml. 

So why was this such a problem in this case? What could we have done to make this more visible to PX4 and QGC so that the change could happen safely?

**Comment by dakejahl** [comment]
> So why was this such a problem in this case? What could we have done to make this more visible to PX4 and QGC so that the change could happen safely?

The enum was re-ordered. In the future if we want to deprecate the use of an enum index we should just label it DEPRECATED but keep it in the enum definition, so as to preserve the ordering.

**Comment by hamishwillee** [comment]
> The enum was re-ordered. In the future if we want to deprecate the use of an enum index we should just label it DEPRECATED but keep it in the enum definition, so as to preserve the ordering.

No. Changing the enum was the right thing to do. We need to be able to change the enums of things in development.xml, including reorders or whatever. Development is just that.

How do we make this more spottable so it doesn't break consumers?

**Comment by dakejahl** [comment]
lol then it will break it for consumers. There's no way around it. There's only one viable solution but you don't like it :shrug:

**Comment by julianoes** [comment]
Things from development.xml should not be released, period.

@MaEtUgR can you give me an example which QGC version against which PX4 version is not compatible? As far as I'm aware standard modes were only released with QGC v4.4.5 and v5.

**Comment by julianoes** [comment]
Related:
https://github.com/PX4/PX4-Autopilot/issues/23940
https://github.com/ArduPilot/pymavlink/pull/240

**Comment by MaEtUgR** [comment]
> example which QGC version against which PX4 version is not compatible?

Timeline:
2022 Feb MAVLink initial pr: https://github.com/mavlink/mavlink/pull/1750
2023 Oct QGC implementation: https://github.com/mavlink/qgroundcontrol/pull/10788
2023 Nov QGC 4.4 release:  https://github.com/mavlink/qgroundcontrol/releases/tag/v4.3.0
2023 Nov PX4 implementation: https://github.com/PX4/PX4-Autopilot/pull/20707
2024 Sep PX4 release 1.15: https://github.com/PX4/PX4-Autopilot/releases/tag/v1.15.0
2024 Dec This breaking change :boom: 
2024 Dec PX4 fixed: https://github.com/PX4/PX4-Autopilot/pull/24011
2025 Jan QGC fixed: https://github.com/mavlink/qgroundcontrol/pull/12276
2025 Jan Message is moved to common: https://github.com/mavlink/mavlink/pull/2161
2025 Jul QGC 5.0: https://github.com/mavlink/qgroundcontrol/releases/tag/v5.0.6
very soon PX4 1.16: https://github.com/PX4/PX4-Autopilot/releases/tag/untagged-c8cdff8d41048ef7d082

So expected incompatibility:
QGC 4.4 - 4.4.5 and PX4 1.16+.
QGC 5+ and PX4 1.15.

In my quick tests:
QGC 5 with PX4 1.15: Return button has no effect. Takeoff shows as empty string.
QGC 4.4.0 with PX4 1.16: works (probably not using standard modes at all because of https://github.com/mavlink/qgroundcontrol/commit/c002ef6f702be1a0ec52914754ab898e3f0e538b)
QGC 5 with PX4 1.16: Return button has no effect. (QGC is not completely fixed yet? :eyes:)

Upstream releases are just the tip of the compatibility iceberg. I’m writing because I’ve seen ground stations misbehaving, showing incorrect modes and reacting wrongly to button presses, even though both sides only ever use implementations from PX4 and QGC that have been available for over a year.

I'm not placing blame, just highlighting that the process isn't working and trying to understand why. People are unknowingly making breaking changes to messages already in use.

If the development dialect can be arbitrarily changed and isn't meant for use, why merge messages into it at all? Why not keep drafts on branches or PRs tied to "unreleased" draft implementations?

Somewhere in Oct/nov 2023 this message should have been moved to the common dialect. Who’s responsible for moving WIP/development messages to the correct place?
I added the orbit message 7 years ago and had to mark it WIP. By that logic, it could still change. But it hasn’t. So who's accountable? Me, the contributors, the release authors of software using it, or MAVLink maintainers? When would it not be considered WIP anymore? Fix: https://github.com/mavlink/mavlink/pull/2316

**Comment by julianoes** [comment]
@MaEtUgR thanks for the list.

Let's go through it:

> 2024 Sep PX4 release 1.15: https://github.com/PX4/PX4-Autopilot/releases/tag/v1.15.0

This is before the API stabilized. Any releases should not use messages from development.xml. PX4 v1.15 is not supposed to support common modes.

- When I build `release/1.15` for 6X, it does not use development.xml, so this one is correct.
- However, [5X has development.xml set by default](https://github.com/PX4/PX4-Autopilot/blob/release/1.15/boards/px4/fmu-v5x/default.px4board#L75), that's wrong and raised by @hamishwillee in https://github.com/PX4/PX4-Autopilot/issues/23940 **before the PX4 release**. 

> So expected incompatibility:
QGC 4.4 - 4.4.5 and PX4 1.16+.
QGC 5+ and PX4 1.15.

What's confusing me is that QGC didn't have support until v5 (only just released last week) and v4.4.5 (released May 28 2025).

I added support in QGC v4.4.5 because I saw [reports](https://discuss.px4.io/t/flight-mode-unknown-1-655360-and-unknown-129-655360/43399/18) that it didn't show them:
- https://github.com/mavlink/qgroundcontrol/commit/4d77f94e5b1d0374964c965fccd2bbe8a9e9aeff
- https://github.com/mavlink/qgroundcontrol/commit/1d6ec1783ecbd36b87a0b6bb92d1d16f95860bd5

So, if I understand it right, this leaves QGC v4.4.5 incompatible with PX4 that should not have supported it in the first place.

QGC v4.4.5 and v5 should be compatible with PX4 v1.16+ once released, if I understand it correctly.

Then, there are main builds of PX4 and QGC but there all bets are off, assuming they use development.xml features which are by definition for development only, and could break.

> If the development dialect can be arbitrarily changed and isn't meant for use, why merge messages into it at all? Why not keep drafts on branches or PRs tied to "unreleased" draft implementations?

The messages need to be somewhere, so that QGC, PX4, AP, etc. can work on implementations. Once implementations exist and have been tested for a bit, that's when they are standardized and moved to common. Without having a place to move them to, it would be a nightmare to implement (or create daily builds from) as you can understand. It might be doable for someone with MAVLink experience but wouldn't be very friendly for someone less used to the message generation, etc.

We used to just merge them into common.xml and put a WIP tag next to it. And guess what, people just ignore that and make releases from it an cementing the messages as is. Basically, we made it harder for people to accidentally use messages by moving them to a development.xml dialect.

Unfortunately, we don't have a compiler warning when development.xml is used. @hamishwillee and I tried to get that in, but it's blocked: https://github.com/ArduPilot/pymavlink/pull/240.

> Somewhere in Oct/nov 2023 this message should have been moved to the common dialect. Who’s responsible for moving WIP/development messages to the correct place?

The stakeholder for the message is responsible, so whoever actually needs it and cares. In this case probably @bkueng should have raised it before the PX4 release, to make sure it makes it in.

I'm sorry, but the people on the MAVLink call, usually @hamishwillee, myself @julianoes, and @auturgy can't be responsible for following up on messages and coordinating between releases. We will try - as @hamishwillee made an issue and raised it. If issues get ignored by PX4 (even Daniel is tagged), what else are we supposed to do? Clearly we should have tagged you @MaEtUgR because you would have cared.

**Comment by dakejahl** [comment]
There is so much language and history wrapped up in the terms "Return" and "RTL". I would consider reverting this.

**Comment by julianoes** [comment]
@dakejahl I don't think we can. If we revert we literally introduce a breaking change between released versions of PX4 and QGC. And I laid out above why we didn't technically break anything but it was the fault of implementations to release with development.xml included.

**Comment by DonLakeFlyer** [comment]
> There is so much language and history wrapped up in the terms "Return" and "RTL". I would consider reverting this.

Things like "Return", "RTL" are used throughout the QGC UI. Calling this "Safe Recovery" as a standard matches non of the existing terminology in QGC. Don't see how this is somehow better.

**Comment by DonLakeFlyer** [comment]
Also as far as I can tell this terminology (which is in the QGC source and affects all firmwares) does not match with ArduPilot which has a normal return and a safe return as a mode of return.

**Comment by hamishwillee** [comment]
@DonLakeFlyer 

In https://github.com/mavlink/mavlink/pull/2191#issuecomment-3243688589 @dakejahl suggested perhaps reverting this based on the historical understanding of the terms. I disagree with this - the selection was intended and delibrate:

1. RTH, RTL, Return are all inaccurate because the vehicle does not necessarily return anywhere - it can go to a rally point, an arbitrary mission landing or whatever.
2. This mode name does not inherit the existing definitions, and that is a good thing. I.e. if it was RTL you would wonder if this meant normal return or smart return, or whatever. It means any mode that goes to a safe recovery.
   In other words, not matching ArduPilot or PX4 definitions is a positive and planned thing.
   
That said, it's a valid argument that RTL or Return mode have historical weight, and perhaps people assume nothing about what they actually do. I'd accept a PR to revert or set some other value, and then try to get broad agreement. Someone else has to do that though - a stakeholder who feels strongly about this.

**Comment by DonLakeFlyer** [comment]
According to mavlink standard modes spec:
`MAV_STANDARD_MODE_SAFE_RECOVERY - Automatic mode that takes vehicle to a predefined safe location via a safe flight path, and may also automatically land the vehicle.`

For a generic spec, this seems wrong to me. Is the previous "go up to a set alt, and then flight back to home" safe? Not really I would say. Given that the spec, provides no capability for an "unsafe" return. If even old "return" is safe then I would contend that has been deemed safe for many years now. Why change?

My opinion is that the concept of "Return" or "RTL" is hugely entrenched terminology across the drone community. Not just PX4 are ArduPilot, all drones. Think of the elephant in the room which is DJI. I don't see how changing such a basic concepts name that has been around for years is a good thing.

If you really think the concept of Return is causing problems then adjust a flight mode name is just the tip of the iceberg:
Screenshot 2025-12-10 at 8 27 10 AM
Screenshot 2025-12-10 at 8 27 29 AM
Screenshot 2025-12-10 at 8 27 37 AM
Screenshot 2025-12-10 at 8 28 07 AM
Screenshot 2025-12-10 at 8 33 43 AM

**Comment by DonLakeFlyer** [comment]
FYI: If folks don't have a huge problem with this, I'd like to change the string for safe recovery mode in QGC back to "Return" so the ui is consistent again with the rest of QGC. I'll let other folks debate the internals of the mavlink spec.

**Comment by DonLakeFlyer** [comment]
And yes, I know I'm being a PITA. These sort of inconsistencies just make me a bit crazy.

**Comment by hamishwillee** [comment]
@DonLakeFlyer You have a point. Of course we could use safe recovery everywhere - I guess this is where we run into standization where all the underlying stuff is still custom. I'll add it to the mav call.

@bkueng To your knowledge, what impact would it have to rename the mode MAV_STANDARD_MODE_SAFE_RECOVERY to MAV_STANDARD_MODE_RETURN ? Obviously its a build time change, but how doable do you think this would be?

**Comment by bkueng** [comment]
> @bkueng To your knowledge, what impact would it have to rename the mode MAV_STANDARD_MODE_SAFE_RECOVERY to MAV_STANDARD_MODE_RETURN ? Obviously its a build time change, but how doable do you think this would be?

I don't think it would be a problem if the enum index stays the same. The actual label/name is set on the GCS side (as per spec from what I remember): https://github.com/mavlink/qgroundcontrol/blob/master/src/Vehicle/StandardModes.cc#L60

**Comment by hamishwillee** [comment]
> > @bkueng To your knowledge, what impact would it have to rename the mode MAV_STANDARD_MODE_SAFE_RECOVERY to MAV_STANDARD_MODE_RETURN ? Obviously its a build time change, but how doable do you think this would be?
> 
> I don't think it would be a problem if the enum index stays the same. The actual label/name is set on the GCS side (as per spec from what I remember): https://github.com/mavlink/qgroundcontrol/blob/master/src/Vehicle/StandardModes.cc#L60

Thanks! 

@DonLakeFlyer Let's see what the dev call says. The GCS could just change the text of the first line of the description if we all agree.

**Comment by MaEtUgR** [comment]
@DonLakeFlyer I largely thought the same but felt I was pain enough in this PR. I assumed a broader audience including Ardupilot, QGC had already agreed this phrasing improved UX. It doesn’t for me. Thanks for calling it out.
